pagopa/io-app
# feat: [IOCOM-823] Chips for Inbox and Archived messages on new DS Messages' Home

⚠️ This PR depends on #5766 ⚠️

## Short description
This PR adds the Chip behaviour to the new DS Messages' Home.
|Inbox|Archived|
|-|-|
|![Simulator Screenshot - iPhone 15 - 2024-05-15 at 14 50 34](https://github.com/pagopa/io-app/assets/5150343/9164fd98-6cff-4087-8f0e-5e10f6144d3a)|![Simulator Screenshot - iPhone 15 - 2024-05-15 at 14 50 37](https://github.com/pagopa/io-app/assets/5150343/59ede441-3b3d-4888-8e80-cb835dcdae89)|

## List of changes proposed in this pull request
New Messages Home is organised as follows:
```
MessagesHome
\_TabNavigationContainer
  \_InboxTab
  \_ArchivedTab
\_ViewPagerContainer
   \_MessageList // Inbox
   \_MessageList // Archive
```
The major challenge is to avoid unnecessary re-renderings in the components included by the`ViewPager`. In order to do so, the shown category (inbox or archive) is stored into redux (`ts/features/messages/store/reducers/allPaginated.ts`), inside the `allPaginated` state (which is not persisted) as a literal which value can be either `inbox` or `archive`. Named `shownCategory`, its value can be retrieved using a selector (`shownMessageCategorySelector`).

The `TabNavigationContainer` component uses such value in order to display the selected tab. On every change, it re-renders itself. On a press event, it dispatches an action to trigger the change into redux and it also instructs the `ViewPager` component to scroll to the desired page. This is necessary since the `ViewPager` does not have a method to show the current page based on a changing value (i.e., it needs to be done programmatically). A guard prevents the event logic to be fired if the already selected Tab is pressed.

The `ViewPagerContainer` shows the Inbox and Archive message lists using the new `MessageList` component (and specialising it by assigning the proper category). Since there are only two pages, both instances of `MessageList` are rendered upon first ViewPagerContainer loading. In order to avoid unnecessary re-renderings when the user scrolls horizontally between the pages, the underlying `ViewPager` uses the `onPageSelected` event to dispatch the redux action that updates the `shownCategory`. This triggers the re-rendering of `TabNavigationContainer`, keeping the UI aligned. This event is triggered only when the page changes (i.e., there is no need for a "same-page" guard like in `TabNavigationContainer`). 

Finally, the logic that triggers the first message loading is placed inside the `TabNavigationContainer`, since it is the only component that can use an `useEffect` hook that triggers every time the screen changes. Loading is fired only if the related message section pot (inbox or archived) is `pot.none`.

As a last note, this PR also moves legacy messages' screens and legacy messages home's components into related `legacy` folders.

## How to test
Enable both new DS and new Messages' Home (from Profile). Check that:
- tap on a Chip switches both Chip and View Pager
- slide on the View Pager switches both View Pager and Chip
- upon first full displaying of each inbox and archive screens, a `reloadAllMessage.request` should be fired (but no more afterwards)
- check that the only component that re-renders itself multiple times is `TabNavigationContainer`
